Add ability to set metadata for object uploads in transfer library.

kahkeng opened the pull request:
Adding the ability for users of TransferClient to set metadata of objects they are uploading. Let me know if there's anything that needs to change with the new public methods I added here.

JonathanHenson commented on aws-cpp-sdk-transfer/include/aws/transfer/UploadFileRequest.h:
```diff
@@ -126,6 +126,11 @@ class AWS_TRANSFER_API UploadFileRequest : public S3FileRequest, public std::ena
     bool HasPassedGetObject() const { return m_getObjectPassed.load(); }
     bool HasPassedListObjects() const { return m_listObjectsPassed.load(); }
 
+    // A map of metadata to store with the object in S3.
```
I'd like for the map to be in the function call instead of in this object. This object does not contain any other s3 object specific data; it is a handle for callbacks and status checks etc...

Also, when you do so, be sure to handle r-values instead of just a copy.

kahkeng replied:
Ok that makes sense, thanks for the review. I was wondering if I should go that route but there are many other request properties available in a PutObjectRequest that are yet to be implemented here (cache control, content encoding, request payer, etc) and this would be an api breaking change each time unless we did function overloading for UploadFile. Would that be acceptable or we're ok changing the function signature each time to add a new property?

JonathanHenson replied:
Two overloads should work. Leave the original, then add one with the map as
const ref, then another as an r-value.
On Jan 12, 2016 10:43 AM, "kahkeng" <notifications@github.com> wrote:

> In aws-cpp-sdk-transfer/include/aws/transfer/UploadFileRequest.h
> <https://github.com/awslabs/aws-sdk-cpp/pull/76#discussion_r49495151>:
>
> > @@ -126,6 +126,11 @@ class AWS_TRANSFER_API UploadFileRequest : public S3FileRequest, public std::ena
> >      bool HasPassedGetObject() const { return m_getObjectPassed.load(); }
> >      bool HasPassedListObjects() const { return m_listObjectsPassed.load(); }
> >
> > +    // A map of metadata to store with the object in S3.
>
> Ok that makes sense, thanks for the review. I was wondering if I should go
> that route but there are many other request properties available in a
> PutObjectRequest that are yet to be implemented here (cache control,
> content encoding, request payer, etc) and this would be an api breaking
> change each time unless we did function overloading for UploadFile. Would
> that be acceptable or we're ok changing the function signature each time to
> add a new property?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/awslabs/aws-sdk-cpp/pull/76/files#r49495151>.
>

JonathanHenson commented:
nice work on this, see my comment above.

kahkeng commented:
Ok I refactored to what I think you were suggesting, let me know if I misunderstood you or if anything needs to change. I understand that we don't want to use the UploadFileRequest object as a container for data. I'm just a little concerned that we will need to have many overloaded methods to transfer the various attributes to the underlying CreateMultipartUploadRequest and PutObjectRequest objects. I'm just doing it for metadata here but there are many more that aren't supported yet.

Thanks!

JonathanHenson commented on aws-cpp-sdk-transfer-tests/TransferTests.cpp:
```diff
@@ -1188,5 +1188,101 @@ TEST_F(TransferTests, ScopeTests)
+
+    Aws::Map<Aws::String, Aws::String> headObjectMetadata = headObjectOutcome.GetResult().GetMetadata();
+    ASSERT_EQ(2u, headObjectMetadata.size());
+    ASSERT_EQ("val1", headObjectMetadata["key1"]);
```
ASSERT_EQ(metadata["key1"], headObjectMetadata["key1"]);

JonathanHenson commented on aws-cpp-sdk-transfer-tests/TransferTests.cpp:
```diff
@@ -1188,5 +1188,101 @@ TEST_F(TransferTests, ScopeTests)
+    Aws::Map<Aws::String, Aws::String> headObjectMetadata = headObjectOutcome.GetResult().GetMetadata();
+    ASSERT_EQ(2u, headObjectMetadata.size());
+    ASSERT_EQ("val1", headObjectMetadata["key1"]);
+    ASSERT_EQ("val2", headObjectMetadata["key2"]);
```
ASSERT_EQ(metadata["key1"], headObjectMetadata["key1"]);

JonathanHenson commented on aws-cpp-sdk-transfer-tests/TransferTests.cpp:
```diff
@@ -1188,5 +1188,101 @@ TEST_F(TransferTests, ScopeTests)
+
+    Aws::Map<Aws::String, Aws::String> headObjectMetadata = headObjectOutcome.GetResult().GetMetadata();
+    ASSERT_EQ(2u, headObjectMetadata.size());
+    ASSERT_EQ("val1", headObjectMetadata["key1"]);
```
ditto here

JonathanHenson commented on aws-cpp-sdk-transfer/include/aws/transfer/UploadFileRequest.h:
```diff
@@ -83,6 +83,22 @@ class AWS_TRANSFER_API UploadFileRequest : public S3FileRequest, public std::ena
+                      const std::shared_ptr<Aws::S3::S3Client>& s3Client,
+                      bool createBucket,
+                      bool doConsistencyChecks);
     ~UploadFileRequest();
```
This one is not your fault, I must have let it slip through a code-review internally, but the destructor needs to be virtual.

JonathanHenson commented on aws-cpp-sdk-transfer/include/aws/transfer/UploadFileRequest.h:
```diff
@@ -255,6 +271,8 @@ class AWS_TRANSFER_API UploadFileRequest : public S3FileRequest, public std::ena
 
     Aws::String m_contentType;
+    Aws::Map<Aws::String, Aws::String> m_metadata;
+    bool m_metadataHasBeenSet;
```
you don't need this, you have m_metadata.size() > 0

JonathanHenson commented on aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp:
```diff
@@ -105,6 +106,34 @@ m_headBucketRetries(0)
+UploadFileRequest(fileName, bucketName, keyName, contentType, s3Client, createBucket, doConsistencyChecks)
+{
+    m_metadataHasBeenSet = true;
+    m_metadata = metadata;
```
Move this to the initialization list, and add a std::forward around it.

JonathanHenson commented on aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp:
```diff
@@ -105,6 +106,34 @@ m_headBucketRetries(0)
+UploadFileRequest(fileName, bucketName, keyName, contentType, s3Client, createBucket, doConsistencyChecks)
+{
+    m_metadataHasBeenSet = true;
+    m_metadata = metadata;
```
initilalization list

JonathanHenson commented on aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp:
```diff
@@ -202,6 +231,10 @@ bool UploadFileRequest::CreateMultipartUpload()
     {
         createMultipartUploadRequest.SetContentType(m_contentType);
     }
+    if (m_metadataHasBeenSet)
```
if(m_metadata.size() > 0)

JonathanHenson commented:
Thanks for your work on this! Just a few thoughts.

kahkeng commented:
Thanks for the feedback, I think I addressed them all in this new commit. Regarding putting metadata into the initialization list, I had to make one of the new constructors the default constructor, otherwise, I'd get a `mem-initializer for 'metadata' follows constructor delegation` error. Also, I tried to read up more about rvalues/std::forward, hope I got its use correct here but let me know if it isn't. Thanks!

JonathanHenson commented:
This looks good. I'm going to pull down and run through the test suite. If everything works correctly, I'll merge the request.

Thanks for your work on this.

kahkeng commented on aws-cpp-sdk-transfer/source/transfer/TransferClient.cpp:
```diff
@@ -74,6 +74,24 @@ std::shared_ptr<UploadFileRequest> TransferClient::UploadFile(const Aws::String&
+
+std::shared_ptr<UploadFileRequest> TransferClient::UploadFile(const Aws::String& fileName, const Aws::String& bucketName, const Aws::String& keyName, const Aws::String& contentType, Aws::Map<Aws::String, Aws::String>&& metadata, bool createBucket, bool doConsistencyChecks)
+{
+    auto request = Aws::MakeShared<UploadFileRequest>(ALLOCATION_TAG, fileName, bucketName, keyName, contentType, metadata, m_s3Client, createBucket, doConsistencyChecks);
```
Actually, to confirm, I think I also need to wrap metadata here with std::forward? If so, I can adjust this patch.

kahkeng commented:
I've added one more instance of `std::forward` in TransferClient when creating the UploadFileRequest object with rvalue metadata, and changed my tests to have `cCreateBucket = true` so they can run individually, and replaced the old commit with this new one. Hope it's all set now.

kahkeng commented:
Updated the diff to use std::move instead of std::forward since I've learnt the former is meant for rvalue references which is the case here, and the latter, although it works here too, seems meant more for universal references. I've also checked via print statements that the rvalue reference constructor of UploadFileRequest gets called when I pass in an rvalue metadata into the TransferClient uploadFile method. Please let me know if it's fine to merge now, thanks!
